compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs

```diff
@@ -856,3 +868,119 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+                                ty::ConstKind::Infer(..) => {
+                                    // Replace not yet inferred const params with their def name.
+                                    self.tcx().mk_const_param(param.index, param.name, c.ty).into()
+                                }
+                                _ => subst,
+                            }
+                        }
+                        _ => subst.super_fold_with(self),
+                    })
+                    .collect();
+                if self.level == 1
+                    || substs.iter().any(|subst| match subst.unpack() {
+                        ty::subst::GenericArgKind::Type(t) => match t.kind() {
+                            ty::Error(_) => false,
+                            _ => true,
+                        },
+                        // Account for `const` params here, otherwise `doesnt_infer.rs`
+                        // shows `_` instead of `Foo<{ _: u32 }>`
+                        ty::subst::GenericArgKind::Const(_) => true,
+                        _ => false,
+                    })
+                {
+                    let substs = self.tcx().intern_substs(&substs[..]);
+                    self.tcx().mk_ty(ty::Adt(def, substs))
+                } else {
+                    self.tcx().ty_error()
+                }
+            }
+            ty::Ref(_, ty, _) => {
+                let ty = self.fold_ty(ty);
+                match ty.kind() {
+                    // Avoid `&_`, these can be safely presented as `_`.
+                    ty::Error(_) => self.tcx().ty_error(),
+                    _ => t.super_fold_with(self),
+                }
+            }
+            // We could account for `()` if we wanted to replace it, but it's assured to be short.
+            ty::Tuple(_)
+            | ty::Slice(_)
+            | ty::RawPtr(_)
+            | ty::FnDef(..)
+            | ty::FnPtr(_)
+            | ty::Opaque(..)
+            | ty::Projection(_)
+            | ty::Never => t.super_fold_with(self),
+            ty::Array(ty, c) => {
+                self.tcx().mk_ty(ty::Array(
+                    self.fold_ty(ty),
+                    match c.val {
+                        ty::ConstKind::Infer(..) => {
+                            // Replace not yet inferred const params with their def name.
+                            self.tcx().mk_const_param(0, Symbol::intern("N"), c.ty).into()
+                        }
+                        _ => c,
+                    },
+                ))
+            }
+            // We don't want to hide type params that haven't been resolved yet.
+            // This would be the type that will be written out with the type param
+            // name in the output.
+            ty::Infer(_) => t,
+            // We don't want to hide the outermost type, only its type params.
+            _ if self.level == 1 => t.super_fold_with(self),
+            // Hide this type
```

Comment:
I'm slightly worried about the handling of `ty::Error`s that come in as part of the input as they become indistinguishable from those that are produced by this `TypeFolder`, but its probably fine…

Reply:
I would think that any naturally ocurring `ty::Error` should be displayed as `_` anyways. Worst case scenario on the next wave of errors the user will be told to fill it in.